[tests] Remove COMPATIBLE_COMPILER_VERSION

Ahajha opened the pull request:
This will be replaced with lockfiles soon. This will be merged after that change.

I have horrifically broken git. Going to scrap my 3 line change and try again.

modularbot committed:
[External] [stdlib] Refactor and optimize `String.splitlines()` (#49663)

[External] [stdlib] Refactor and optimize `String.splitlines()`

Refactor and optimize `String.splitlines()`, add method `isnewline()` to
`StringSlice` only since it is not a Python API but very necessary when
going over string views to get better perf.

Added a method `_to_string_list()` which is very fast in allocating list
of strings from either `List[Span[Byte]]` or `List[StringSlice]`.

ORIGINAL_AUTHOR=martinvuyk
<110240700+martinvuyk@users.noreply.github.com>
PUBLIC_PR_LINK=modularml/mojo#3697

Co-authored-by: martinvuyk <110240700+martinvuyk@users.noreply.github.com>
Closes modularml/mojo#3697
MODULAR_ORIG_COMMIT_REV_ID: cb7604feb033289021057eaf14f5a2a2ee62bd84

gabrieldemarmiesse committed:
[External] [stdlib] Use SIMD to make `b64encode` 4.7x faster (#49573)

[External] [stdlib] Use SIMD to make `b64encode` 4.7x faster

## Dependencies

The following PR should be merged first:
* https://github.com/modularml/mojo/pull/3397

## Description of the changes

`b64encode` is the function that encode bytes to base 64. Base 64
encoding is massively used across the industry, being to write secrets
as text or to send data across the internet.

Since it's going to be used a lot, we should make sure it is fast. As
such, this PR provides a new implementation of `b64encode` around 5
times faster than the current one.

This implementation was taken from the following papers:

Wojciech Muła, Daniel Lemire, Base64 encoding and decoding at almost the
speed of a memory copy, Software: Practice and Experience 50 (2), 2020.
https://arxiv.org/abs/1910.05109
Wojciech Muła, Daniel Lemire, Faster Base64 Encoding and Decoding using
AVX2
Instructions, ACM Transactions on the Web 12 (3), 2018.
https://arxiv.org/abs/1704.00605

Note that there are substancial differences between the papers and this
implementation. There are two reasons for this:
* We want to avoid using assembly/llvm intrinsics directly and try to
use the functions provided by the stdlib
* We want to keep the complexity low, so we don't make a slightly
different algorithm for each simd sizes and each cpu architecture.

In a nutshell, we decide on a simd size, let's say 32. So at each
iteration, we load 32 bytes, reshuffle the 24 first bytes, convert them
to base 64, it then becomes 32 bytes, and then we store those 32 bytes
in the output buffer.

We have a final iteration with the last incomplete chunks where we
shouldn't load everything at once, otherwise we would get out of bounds
errors. We then use partial loads and store and masking, but the main
SIMD algorithm is used.

The reasons for the speedups are simlar to the ones provided in
https://github.com/modularml/mojo/pull/3401

## API changes

The existing api is
```mojo
fn b64encode(str: String) -> String:
```
and has several limitations:
1) The input of the function is raw bytes. It doesn't have to represent
text. Requirering the user to provide a `String` forces the user to
handle null termination on its bytes and whatever other requirement
`String` might have to use bytes.
2) It is not possible to write the produced bytes in an existing buffer.
3) It is hard to benchmark as the signature implies that the function
allocates memory on the heap.
4) It supposes that the input value owns the underlying data, meaning
that it's not possible to use the function if the data is not owned.
`Span` would be a better choice here.

We keep in this PR the existing signature for backward compatibility and
add new overloads. Now the signatures are:
```mojo
fn b64encode(input_bytes: List[UInt8, _], inout result: List[UInt8, _])
fn b64encode(input_bytes: List[UInt8, _]) -> String
fn b64encode(input_string: String) -> String
```

Note that it could be further improved in future PRs as currently `Span`
is not easy to use but would be a right fit for the input value. We
could also in the future remove `fn b64encode(input_string: String) ->
String`.

Note that the python api takes `bytes` as input and returns `bytes`.

## Benchmarking

Benchmarking is harder than usual here because the base function does
memory allocation. To avoid having the alloc in the benchmark, we must
modify the original function to add the overloads described above. In
this case we can benchmark and on my system

```
WSL2 windows 11
Intel(R) Core(TM) i7-10700KF CPU @ 3.80GHz
Base speed:	3,80 GHz
Sockets:	1
Cores:	8
Logical processors:	16
Virtualization:	Enabled
L1 cache:	512 KB
L2 cache:	2,0 MB
L3 cache:	16,0 MB
```
We get around 5x speedup.

I don't provide the benchmark script here because it won't work out of
the box (see the issue mentionned above), but if that's really necessary
to get this merged, I'll provide the diff + the benchmark script.

## Future work

As said before, this PR is not an exact re-implementation of the papers
and the state of the art implementation that comes with it, the
[simdutf](https://github.com/simdutf/simdutf) library.

This is to keep this implementation simple and portable as it will work
on any CPU that has an simd size of at least 4 bytes, and below or equal
64 bytes.

In future PRs, we could provide futher speedups by using simd algorithms
that are specific to each architecture. This will greatly increase the
complexity of the code. I'll leave this decision to the maintainers.

We can also re-write `b64decode` using simd and it's also expected that
we'll get speedups. This can be the topic of another PR too.

Co-authored-by: Gabriel de Marmiesse <gabrieldemarmiesse@gmail.com>
Closes modularml/mojo#3443
MODULAR_ORIG_COMMIT_REV_ID: 0cd01a091ba8cfdaac49dcf43280de22d9c8b299

Ahajha committed:
Remove COMPATIBLE_COMPILER_VERSION

Signed-off-by: Alex Trotta <atrotta@modular.com>

Ahajha committed:
[tests] Remove COMPATIBLE_COMPILER_VERSION check

MODULAR_ORIG_COMMIT_REV_ID: e37f767f0c6c21f2c562bd2d9372c909a5af42f5

JoeLoser reviewed (approved):
LGTM as long as we have the necessary functionality internally to still automatically test the pre-releases in our CI internally.  This was just the minimal thing we needed when we initially open sourced.

JoeLoser committed:
[stdlib] Fix `run-benchmarks` lit config

cb7604feb033289021057eaf14f5a2a2ee62bd84 broke the OSS benchmarks when
not run internally. This is due to a simple type-checking issue:
`pre_built_packages_path` is a string type which won't have a
`resolve()` function. Explicitly construct it to a `Path` type instead
which has the needed `resolve()` functionality.

MODULAR_ORIG_COMMIT_REV_ID: a48a48db3428d7fc431a9f5403ad813b59484e79

judyheflin committed:
[mojo-stdlib][docs] Updates to Mojo type manual (StringLiteral, UInt,
and Tuple)

Added clarification/updates for StringLiteral and UInt. Added a
subsection and example for Tuple.

MODULAR_ORIG_COMMIT_REV_ID: 8649f79626b0583f00ed13a1f9cdd09034be67a0

Ahajha committed:
Only look at the lockfiles in CI

Signed-off-by: Alex Trotta <atrotta@modular.com>

soraros committed:
[External] [stdlib] Make `StringRef` ctor from `UnsafePointer[Byte]` keyword only (#49581)

[External] [stdlib] Make `StringRef` ctor from `UnsafePointer[Byte]`
keyword only

This issues was reported by `@aurelian`(@diocletiann) on Discord.
This prevents accidental `UnsafePointer[Byte]` to `String` implicit
conversion through `String` ctor from `StringRef`.

Make this explicit by explicitly constructing the `StringRef` with the
keyword-only constructor.

Co-authored-by: soraros <soraros@users.noreply.github.com>
Closes modularml/mojo#3698
MODULAR_ORIG_COMMIT_REV_ID: 44c1e061db833fff3cf261f2b6eabfcbef6e1591

bgreni committed:
[External] [stdlib] Add benchmark for `lcm` and `gcd` (#41811)

[External] [stdlib] Add benchmark for `lcm` and `gcd`

Add benchmarks for `lcm` and `gcd` functions.

Co-authored-by: bgreni <42788181+bgreni@users.noreply.github.com>
Closes modularml/mojo#3041
MODULAR_ORIG_COMMIT_REV_ID: 3fe9e4a6321832878a49ad79d90593ef2969076b

Ahajha committed:
[CI] Un-ignore lockfiles again

MODULAR_ORIG_COMMIT_REV_ID: b510830590117b3eb28db46074cfbe75ad7989bd

carolinefrasca committed:
[Docs] Remove mentions of (now closed) GitHub Discussions from the MAX
Graph tutorial and a few README files.

MODULAR_ORIG_COMMIT_REV_ID: 8d212fe410e8fa53695a627a3f3444101025c2a7

modularbot committed:
Update lockfiles to point to latest nightly version: 24.6.0.dev2024103019

abduld committed:
[******][GPU] Fix the SM90 arch detection

MODULAR_ORIG_COMMIT_REV_ID: c877b952dd7136fe53e11990bc207eed7539ecd7

modularbot committed:
Update lockfiles to point to latest nightly version: 24.6.0.dev2024103105

Ahajha committed:
Move --frozen to the correct place
